tda/matrix.py

```diff
@@ -0,0 +1,289 @@
+
+        Notes:
+        ------
+        The method uses modulo 2 addition of entries.
+
+        """
+
+        row_to_change = self.get_row(index_row_to_change)
+        row_to_add = self.get_row(index_row_to_add)
+        summed_row = []
+        for i in range(self.num_cols):
+            summed_entries = (row_to_change[i] + row_to_add[i]) % 2
+            summed_row.append(summed_entries)
+        self.change_row(index_row_to_change, summed_row)
+
+    def add_cols(self, index_col_to_change, index_col_to_add):
+        """
+
+        Add one col to another.
+
+        Parameters:
+        -----------
+        index_col_to_change: int
+            Index of changed col.
+        index_col_to_add: int
+            Index of added col to another.
+
+        Notes:
+        ------
+        The method uses modulo 2 addition of entries.
+
+        """
+
+        col_to_change = self.get_col(index_col_to_change)
+        col_to_add = self.get_col(index_col_to_add)
+        summed_col = []
+        for i in range(self.num_rows):
+            summed_entries = (col_to_change[i] + col_to_add[i]) % 2
+            summed_col.append(summed_entries)
+        self.change_row(index_col_to_change, summed_col)
+
+    def non_zero(self, entry_index):
+        """
+
+        Arranges rows and cols so that specified entry on diagonal is non zero.
+
+        Parameters:
+        -----------
+        entry_index: int
+            Index of a row (and column) that contains the entry.
+
+        Output:
+        -------
+        value_check: bool
+            Boolean that tells whether the entry is not zero.
+
+        Notes:
+        ------
+        Intended to use only in smith_normal_form method due to algorithm
+        optimizations.
+
+        It will move a row, if non zero entry is on the same column as the
+        desired place.
+
+        It will move a columns, if non zero entry is right of
+        the desired place.
+
+        """
+
+        zero_row = True
+        for i in range(self.num_cols):
+            if self.entries[entry_index][i] != 0:
+                zero_row = False
+        if zero_row is True:
+            return False
+
+        if self.entries[entry_index][entry_index] != 0:
+            return True
+
+        for row_index in chain(range(entry_index),
+                               range(entry_index+1, self.num_rows)):
+
+            if self.entries[row_index][entry_index] != 0:
+                copied_row = copy(self.get_row(row_index))
+                copied_entry_row = copy(self.get_row(entry_index))
+
+                self.change_row(entry_index, copied_row)
+                self.change_row(row_index, copied_entry_row)
+
+                return True
+
+        for col_index in chain(range(entry_index),
+                               range(entry_index+1, self.num_cols)):
+            if self.entries[entry_index][col_index] != 0:
+                copied_col = copy(self.get_col(col_index))
```

Comment:
Can we implement this without making copies? The cost of copying a list is O(n).